common/dagmcmetadata.cpp

```diff
@@ -0,0 +1,561 @@
+  } else if ( property == "density" ) {
+    value = volume_density_data_eh[eh];
+  } else if ( property == "importance" ) {
+    value = volume_importance_data_eh[eh];
+  } else if ( property == "tally" ) {
+    value = tally_data_eh[eh];
+  } else {
+    std::cout << "Not a valid property for volumes" << std::endl;
+  }
+  return value;
+}
+
+// overloaded get_volume_property for indices and id's'
+std::string dagmcMetaData::get_volume_property(std::string property, int vol, bool idx)
+{
+  // if this is an index query
+  moab::EntityHandle eh;
+  if ( idx == true) {
+    eh = DAG->entity_by_index( 3, vol );
+  } else {
+    eh = DAG->entity_by_id( 3, vol );
+  }
+  return get_volume_property(property,eh);
+}
+
+// Get a given property on a surface
+std::string dagmcMetaData::get_surface_property(std::string property, moab::EntityHandle eh)
+{
+  std::string value = "";
+  if (property == "boundary" ) {
+    value = surface_boundary_data_eh[eh];
+  } else if (property == "tally") {
+    value = tally_data_eh[eh];
+  } else {
+    std::cout << "Not a valid property for surfaces" << std::endl;
+  }
+  return value;
+}
+
+// overloaded get_surface_property for indices and id's'
+std::string dagmcMetaData::get_surface_property(std::string property, int vol, bool idx)
+{
+  // if this is an index query
+  moab::EntityHandle eh;
+  if ( idx == true) {
+    eh = DAG->entity_by_index( 2, vol );
+  } else {
+    eh = DAG->entity_by_id( 2, vol );
+  }
+  return get_surface_property(property,eh);
+}
+
+// parse the material data
+void dagmcMetaData::parse_material_data()
+{
+  std::map<moab::EntityHandle,std::vector<std::string> > material_assignments;
+  material_assignments = get_property_assignments("mat",3,":/",false);
+  std::map<moab::EntityHandle,std::vector<std::string> > density_assignments;
+  density_assignments = get_property_assignments("rho",3,":",false);
+
+  int num_cells = DAG->num_entities( 3 );
+
+  std::vector<std::string> material_props;
+  std::vector<std::string> density_props;
+
+  std::string implicit_complement_material = "";
+  std::string implicit_complement_density = "";
+
+  // loop over all cells
+  for( int i = 1; i <= num_cells; ++i ) {
+
+    int cellid = DAG->id_by_index( 3, i );
+    moab::EntityHandle eh = DAG->entity_by_index( 3, i );
+
+    material_props = material_assignments[eh];
+    density_props = density_assignments[eh];
+
+    // this is actually ok for a single volume, one that has the _comp tag at the end of it
+    if( material_props.size() > 1 ) {
+      // search the props for _comp
+      std::size_t comp_found;
+      int position;
+      for ( int j = 0 ; j < material_props.size() ; j++ ) {
+        comp_found = material_props[j].find("_comp");
+        if ( comp_found != std::string::npos) {
+          position = j;
+          break;
+        }
+      }
+      // if there is no material property - not failure for impl_comp
+      if(material_props[0] == "" && !(DAG->is_implicit_complement(eh))) {
```

Comment:
I just tested it and if you have a volume with no material, it doesn't give an error and eventually segfaults when MCNP tries to load the lcad file. If you move it above this block, then you get the expected behavior; i.e. `more than one material for volume with id`